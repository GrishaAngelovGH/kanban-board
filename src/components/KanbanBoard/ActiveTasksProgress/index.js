import ActiveTasksProgress from "./ActiveTasksProgress"

export default ActiveTasksProgress