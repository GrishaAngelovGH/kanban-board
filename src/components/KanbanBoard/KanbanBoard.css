.kanban-board {
  height: 100%;
  background-size: cover;
}

.kanban-board .toast-container {
  position: fixed;
}

.react-calendar {
  position: fixed;
  right: 4%;
  z-index: 10;
  border: 5px solid #e3e3e3;
  border-bottom-left-radius: 5px;
  border-bottom-right-radius: 5px;
}

.modal-open > .modal-backdrop {
  --bs-backdrop-bg: transparent;
  --bs-backdrop-opacity: 1;
  backdrop-filter: blur(100px);
}