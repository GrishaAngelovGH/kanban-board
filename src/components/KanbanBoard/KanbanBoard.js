import { useState, useEffect } from "react"

import ButtonPanel from "./ButtonPanel"
import Column from "./Column"
import ColumnModal from "./ColumnModal"

import boardRepository from "persistent/persistentKanbanBoardRepository"

const KanbanBoard = () => {
  const [showColumnModal, setShowColumnModal] = useState(false)
  const [columns, setColumns] = useState([])

  const toggleColumnModal = () => {
    setShowColumnModal(!showColumnModal)
  }

  const handleConfirm = (title, description) => {
    boardRepository.createColumn(title, description)
    setShowColumnModal(!showColumnModal)
  }

  const handleDeleteColumn = title => {
    boardRepository.deleteColumn(title)
    setColumns(boardRepository.getColumns())
  }

  const handleClearBoard = () => {
    boardRepository.deleteAllColumns()
    setColumns(boardRepository.getColumns())
  }

  useEffect(() => {
    setColumns(boardRepository.getColumns())
  }, [showColumnModal])

  return (
    <div className="row">
      <div className="col-md-12">

        <div className="row p-5">
          {
            columns.map((v, i) => (
              <Column
                key={i}
                title={v.title}
                description={v.description}
                onDelete={handleDeleteColumn}
              />
            ))
          }
        </div>

        {
          showColumnModal && (
            <ColumnModal
              show={showColumnModal}
              onClose={toggleColumnModal}
              onConfirm={handleConfirm}
            />
          )
        }

        <ButtonPanel
          onColumnButtonClick={toggleColumnModal}
          onClearBoardButtonClick={handleClearBoard}
        />
      </div>
    </div>
  )
}

export default KanbanBoard