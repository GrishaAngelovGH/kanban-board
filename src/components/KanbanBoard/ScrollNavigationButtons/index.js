import ScrollNavigationButtons from "./ScrollNavigationButtons"

export default ScrollNavigationButtons