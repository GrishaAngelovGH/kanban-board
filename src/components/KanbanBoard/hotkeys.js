import hotkeys from "hotkeys-js"

const configureHotkeys = handlers => {
  hotkeys("ctrl+g", (event, handler) => {
    event.preventDefault()

    const keys = {
      "ctrl+g": handlers.ctrlPlusG
    }

    keys[handler.key]()
  })
}

export default configureHotkeys