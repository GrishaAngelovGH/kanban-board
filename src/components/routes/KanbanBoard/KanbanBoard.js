import { useState, useEffect, useCallback } from "react"

import Layout from "components/Layout"
import Header from "components/Header"
import KanbanBoard from "components/KanbanBoard"
import Users from "components/Users"
import Settings from "components/Settings"

import boardRepository from "persistent/persistentKanbanBoardRepository"
import settingsRepository from "persistent/persistentSettingsRepository"
import history from "persistent/history"

const KanbanBoardRoute = ({ showToastWithMessage }) => {
  const [showSettings, setShowSettings] = useState(false)
  const [showCalendar, setShowCalendar] = useState(false)
  const [showUsers, setShowUsers] = useState(false)
  const [downloadJsonHref, setDownloadJsonHref] = useState("")
  const [showUploadBoardModal, setShowUploadBoardModal] = useState(false)

  const toggleSettings = () => {
    setShowSettings(!showSettings)
  }

  const toggleCalendar = () => {
    setShowCalendar(!showCalendar)
  }

  const toggleUsers = () => {
    setShowUsers(!showUsers)
  }

  const prepareJsonFileDownload = useCallback(() => {
    const title = boardRepository.getBoardTitle()
    const columns = boardRepository.getStringifiedColumns()
    const historyColumns = history.getStringifiedColumns()

    const jsonString = `{"title": "${title}", "columns": ${columns}, "history": ${historyColumns}}`
    const blob = new Blob([jsonString], { type: "application/json" })
    setDownloadJsonHref(URL.createObjectURL(blob))
  }, [])

  const toggleUploadKanbanBoardModal = () => {
    setShowUploadBoardModal(!showUploadBoardModal)
  }

  const handleConfirmSettings = settings => {
    settingsRepository.saveSettings(settings)
    setShowSettings(false)
  }

  useEffect(prepareJsonFileDownload, [prepareJsonFileDownload])

  const disabledExport = !boardRepository.getColumns().length

  const headerHandlers = {
    onCalendarClick: toggleCalendar,
    onImportKanbanBoardClick: toggleUploadKanbanBoardModal,
    onSettingsClick: toggleSettings,
    onUsersClick: toggleUsers,
    onUpdate: prepareJsonFileDownload
  }

  return (
    <div className="container-fluid">
      <Layout
        header={
          <Header
            downloadJsonHref={downloadJsonHref}
            disabledExport={disabledExport}
            handlers={headerHandlers}
          />
        }
        body={
          <KanbanBoard
            showCalendar={showCalendar}
            showUploadBoardModal={showUploadBoardModal}
            onUpdate={prepareJsonFileDownload}
            onToggleUploadKanbanBoardModal={toggleUploadKanbanBoardModal}
            showToastWithMessage={showToastWithMessage}
          />
        }
      />

      <Users
        show={showUsers}
        onClose={toggleUsers}
        onUpdate={prepareJsonFileDownload}
      />

      <Settings
        show={showSettings}
        onClose={toggleSettings}
        onConfirm={handleConfirmSettings}
      />
    </div>
  );
}

export default KanbanBoardRoute
