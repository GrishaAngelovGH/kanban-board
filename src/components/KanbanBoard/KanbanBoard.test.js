import { render } from "@testing-library/react"

import KanbanBoard from "./KanbanBoard"

test("should render KanbanBoard component", () => {
  const onUpdate = jest.fn()

  const view = render(<KanbanBoard onUpdate={onUpdate} />)

  expect(view).toMatchSnapshot()
})