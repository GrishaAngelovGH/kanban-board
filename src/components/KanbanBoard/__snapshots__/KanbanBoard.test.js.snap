// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render KanbanBoard component 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="row"
      >
        <div
          class="col-md-12"
        >
          <div
            class="row p-5"
          />
          <div
            class="row border-top border-secondary-subtle fixed-bottom p-2 justify-content-end text-end button-panel"
          >
            <div
              class="col col-md-4 col-lg-3"
            >
              <div
                class="btn-group"
                role="group"
              >
                <button
                  class="btn btn-primary"
                  type="button"
                >
                  New Column
                </button>
                <button
                  class="btn btn-danger"
                  type="button"
                >
                  Clear Board
                </button>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="row"
    >
      <div
        class="col-md-12"
      >
        <div
          class="row p-5"
        />
        <div
          class="row border-top border-secondary-subtle fixed-bottom p-2 justify-content-end text-end button-panel"
        >
          <div
            class="col col-md-4 col-lg-3"
          >
            <div
              class="btn-group"
              role="group"
            >
              <button
                class="btn btn-primary"
                type="button"
              >
                New Column
              </button>
              <button
                class="btn btn-danger"
                type="button"
              >
                Clear Board
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
